### ISSUE TITLE
None becomes "" when assigned to variable

### ISSUE THREAD

**Issue by wknapik** [ansible/ansible#41392]
##### SUMMARY
None becomes `""` when assigned to variable.

A comparison between a variable and what was assigned to it will return `false`.

Ran into this while using `lookup('file', '/nonexistent', errors='ignore')` - had to change a `when:` condition depending on whether comparing the lookup, or a variable with the lookup result.

##### ISSUE TYPE
Bug Report

##### COMPONENT NAME

core

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/foo/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.5 (default, May 11 2018, 04:00:52) [GCC 8.1.0]
```

##### OS / ENVIRONMENT

N/A

##### STEPS TO REPRODUCE

```yaml
- set_fact:
    foo: "{{ None }}"
- debug:
    msg: "{{ foo == None }}"
- debug:
    msg: "{{ foo == '' }}"
```

##### EXPECTED RESULTS
`true`, `false`

##### ACTUAL RESULTS
`false`, `true`

```
TASK [debug] **********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": false
}

TASK [debug] **********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": true
}
```

**Comment by sivel** [comment]
This is specifically implemented functionality of our jinja2 implementation:

https://github.com/ansible/ansible/blob/a9e53cdb68cfb80a69297a8ab6266d0654d9b6a1/lib/ansible/template/__init__.py#L610-L614

At this point I'm not sure it would be feasible to change this implementation.

**Comment by wknapik** [comment]
Well, that's pretty terrible ;]
I mean it's `foo = bar; foo != bar`. That can't be considered correct, right ?;]

**Comment by sivel** [comment]
It's possible, that we could bypass this if utilizing native jinja2 types: 

https://github.com/ansible/ansible/blob/edb1735ca07059a3b644867045f70f9f4ee63edd/lib/ansible/config/base.yml#L697-L706

@jctanner @mkrizek do you guys have an opinion on whether we should skip the `_finalize` functionality when using native types?

**Comment by bcoca** [comment]
The other problem is that None means nothing in 'YAML'  so it would be translated into the string 'None', for yaml `~` is the Null/None token.

**Comment by sivel** [comment]
> The other problem is that None means nothing in 'YAML' so it would be translated into the string 'None', for yaml ~ is the Null/None token.

Yes, I think it would be best if we only allowed it with native types as a result of this.  Unsure if other changes would have to happen to permit this.

**Comment by wknapik** [comment]
Sweet!

Are there any other side-effects of using `jinja2_native = true` ? Anything that could potentially cause trouble ? The docs don't really say...

**Comment by sivel** [comment]
@wknapik it is new functionality, that while tested, hasn't been used in the wild much.

That functionality was added to fix some issues with our historical integration with jinja2.

jinja2 has historically only returned strings, so the effect that had when returning something like `None` was that jinja2 actually returned a repr of `None` which is `'None'`.  We then have a function (`safe_eval`) that will attempt to safely eval the return from jinja2 to turn those reprs back into python data structures.

It is possible that you or other users had run into a situation where `safe_eval` didn't do the right thing, and may have worked around it in some manner that could potentially be incompatible with the change to use `jinja2_native`.

In the end, try it and see.  I will however mention that `jinja2_native` is not in any released version, and will be included in the 2.7 release which is currently `devel` at this time.

**Comment by wknapik** [comment]
@sivel thanks a lot for the explanation!
